akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -0,0 +1,985 @@
+    if (unwiredOuts == 0) {
+      var traversal: Traversal = finalSteps
+      var remaining = reverseBuildSteps
+      while (remaining.nonEmpty) {
+        remaining.head match {
+          case key: BuilderKey ⇒
+            // At this point all the builders we have are completed and we can finally build our traversal
+            traversal = pendingBuilders(key).traversal.concat(traversal)
+          case AppendTraversal(toAppend) ⇒
+            traversal = toAppend.concat(traversal)
+        }
+        remaining = remaining.tail
+      }
+
+      val finalTraversal = if (islandTag == null) traversal else EnterIsland(islandTag, traversal)
+
+      // The CompleteTraversalBuilder only keeps the minimum amount of necessary information that is needed for it
+      // to be embedded in a larger graph, making partial graph reuse much more efficient.
+      CompletedTraversalBuilder(
+        traversalSoFar = finalTraversal,
+        inSlots,
+        inOffsets,
+        attributes
+      )
+    } else this
+  }
+
+  /**
+   * Assign an output port a relative slot (relative to the base input slot of its module, see [[MaterializeAtomic]])
+   */
+  override def assign(out: OutPort, relativeSlot: Int): TraversalBuilder = {
+    // Which module out belongs to (indirection via BuilderKey and pendingBuilders)
+    val builderKey = outOwners(out)
+    val submodule = pendingBuilders(builderKey)
+
+    // Do the assignment in the submodule
+    val result = submodule.assign(out.mappedTo, relativeSlot)
+    val wired = if (result.isTraversalComplete) {
+      // Remove the builder (and associated data).
+      // We can't simply append its Traversal as there might be uncompleted builders that come earlier in the
+      // final traversal (remember, input ports are assigned in traversal order of modules, and the inOffsets
+      // and inBaseOffseForOut Maps are updated when adding a module; we must respect addition order).
+
+      copy(
+        inBaseOffsetForOut = inBaseOffsetForOut - out,
+        outOwners = outOwners - out,
+        // TODO Optimize Map access
+        pendingBuilders = pendingBuilders.updated(builderKey, result),
+        // pendingBuilders = pendingBuilders - builderKey,
+        unwiredOuts = unwiredOuts - 1
+      )
+    } else {
+      // Update structures with result
+      copy(
+        inBaseOffsetForOut = inBaseOffsetForOut - out,
+        unwiredOuts = unwiredOuts - 1,
+        pendingBuilders = pendingBuilders.updated(builderKey, result)
+      )
+    }
+
+    // If we have no more unconnected outputs, we can finally build the Traversal and shed most of the auxiliary data.
+    wired.completeIfPossible
+  }
+
+  // Requires that a remapped Shape's ports contain the same ID as their target ports!
+  def add[A, B, C](submodule: TraversalBuilder, shape: Shape, combineMat: (A, B) ⇒ C): TraversalBuilder = {
+    val builderKey = new BuilderKey
+
+    val newBuildSteps =
+      if (combineMat == Keep.left) {
```

Comment:
We convert those earlier in the javadsl so we don't have to care about javadsl.
Perhaps worth a comment.